visualizer/draw/area-chart.js

```diff
@@ -387,23 +451,69 @@ class AreaChart extends HtmlContent {
     this.d3SliceHighlight.style('top', margins.top + 'px')
   }
 
-  drawFiltering () {
-    this.d3AreaPaths.classed('filtered', d => d.key.split('_')[1] === 'absent' || (this.layoutNode && this.layoutNode.id !== extractLayoutNodeId(d.key)))
-  }
-
   draw () {
     super.draw()
 
     if (this.layoutNodeToApply) {
       this.layoutNode = this.layoutNodeToApply
       this.layoutNodeToApply = null
-      this.drawFiltering()
+      // this.drawFiltering()
+      this.drawToCanvas(true)
     }
     if (this.widthToApply && this.widthToApply !== this.width) {
       this.drawPathsToFit(this.widthToApply)
+      this.drawToCanvas()
     }
     if (!this.isInHoverBox) this.widthToApply = null
   }
+
+  drawToCanvas (filtered = false) {
+    this.canvasContext.clearRect(0, 0, this.d3CanvasPlotArea.attr('width'), this.d3CanvasPlotArea.attr('height'))
+    this.dataContainer.selectAll('custom.area').each((d, i, nodes) => {
+      const node = d3.select(nodes[i])
+      const { fillColour, opacity, blendMode } = this.getCanvasAreaStyles(node.attr('cssId'), node.attr('isEven'), filtered && node.attr('isFiltered'))
+      this.canvasContext.globaleCompositeOperation = blendMode
+      this.canvasContext.globalAlpha = opacity
+      this.canvasContext.beginPath()
+      this.areaMaker.context(this.canvasContext)(d)
+      this.canvasContext.fillStyle = fillColour
+      this.canvasContext.fill()
+    })
+    this.drawToMouseClickCanvas()
+  }
+
+  getHiddenMouseValues (event) {
+    const { offsetX, offsetY } = event
+    const col = this.hiddenMouseMapperCanvasContext.getImageData(offsetX, offsetY, 1, 1).data
+    return this.colourToNode['rgb(' + col[0] + ',' + col[1] + ',' + col[2] + ')']
```

Comment:
Very creative approach, but on complex area charts (e.g. the 1005 uploads previously linked to), mouse movements seem to sometimes fall between the cracks and not register anything - presumably because of anti-aliasing between fine lines causing the colour at a pixel to not be an exact match? (or potentially a mismatch) 

It also almost doubles the processing time for each redraw. So if we can make the second canvas redundant using a different approach, it should be more reliable, maintainable, and maybe a ~60% perf improvement on heavy profiles compared to current master:

![image](https://user-images.githubusercontent.com/29628323/53975642-35801200-40fd-11e9-8971-7f533247aef1.png)

I'd favour an approach that references the X / Y position back to the data. `showSlice()` already does something like this to get the data index along the x axis for the "XXX pending async operations" tooltip message, and Clinic Flame's dependency `d3-fg uses a similar approach in [its `getNodeAt` function](https://github.com/davidmarkclements/d3-fg/blob/276cc403351766f309ad049bd46790afcccd5b86/index.js#L689) (though that's not necessarily directly applicable because it's based on d3.partition instead of area).

Figuring out which node is at the corresponding point on the Y axis might be trickier than getting the data array index for the X axis because of the data stacking, but I think you could do it if you access the `stackedData` array (either via D3 or store it as `this.stackedData`). This nested array is structured like `stackedData[indexOfNode][indexOfXPosition] = [valueAtY0, valueAtY1]`. 

So maybe, use the y scale to get a number (say, `cursorYValue`) corresponding to the event's adjusted y offset, then loop through the nodes until you find one where `stackedData[i][xIndex][0] < cursorYValue && stackedData[i][xIndex][0] >= cursorYValue`, then `stackedData[i].key` is the one to highlight / select.

Reply:
great feedback - and I agree - this should be improved where possible.  Thanks for the leads o a more calculation based approach  - I will investigate.